scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,175 @@
+"""
+Disjoint set data structure
+"""
+import collections
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_elements : int
+        The number of elements in the set.
+    n_subsets : int
+        The number of subsets.
+
+    Methods
+    -------
+    add
+    merge
+    connected
+    __getitem__
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    or *merge-find* data structure. The *find* operation (implemented in
+    `__getitem__`)* implements the *path halving* variant. The *merge* method
+    implements the *merge by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> disjoint_set = DisjointSet([1, 2, 3, 'a', 'b'])
+
+    Merge some subsets:
+
+    >>> disjoint_set.merge(1, 2)
+    True
+    >>> disjoint_set.merge(3, 'a')
```

Comment:
I changed the behaviour so that `disjoint_set.merge(3, 10)` will now raise `KeyError: 10`.
Elements must now be explicity added in the `__init__` or `add` methods.